### ISSUE TITLE
Misc. bug: CPU backend: llama_batch.embd (embedding input) decoding is non-deterministic, while the token path is bitwise deterministic

### ISSUE THREAD

**Issue by SgfKrc** [ggml-org/llama.cpp#28963]
### Name and Version

$ ./llama-cli --version
version: 0.1.0-dev (build 10436, commit 6f04274cc)
built with GNU 15.2.0 for Windows AMD64

### Operating systems

Windows

### Which llama.cpp modules do you know to be affected?

libllama (core library)

### Command line

```shell
probe.exe model.gguf embd5.f32 repeat 6 1
probe.exe model.gguf embd5.f32 tokens 6 1
```

### Problem description & steps to reproduce

# Title
CPU backend: `llama_batch.embd` (embedding input) decoding is non-deterministic, while the token path is bitwise deterministic

# Summary
On the CPU backend, repeated identical decodes through the **embedding input path**
(`llama_batch.embd` set, `ubatch->token == NULL`) produce **different logits** — in the same
process, same context, same input, after `llama_memory_clear(mem, true)` between decodes.
The **token path** (`llama_batch.token`) is **bitwise identical** under exactly the same
conditions. `argmax` stays stable, so the practical impact is on *reproducibility*, not on
greedy generation.

# Environment
- commit: `6f04274cc145cf76aafc46eb8c1ea054ca221368`
- build: `cmake -B build-cpu -G "MinGW Makefiles" -DGGML_CUDA=OFF -DLLAMA_BUILD_TESTS=OFF` (MSYS2 UCRT64 g++)
- runtime: Windows amd64, CPU backend, `n_gpu_layers = 0`, `n_threads = n_threads_batch = 1`
- model: Qwen3.5-2B (`qwen35`: hybrid SSM + full attention every 4 layers), f16 GGUF
- ctx: `n_ctx=512`, `n_batch=n_ubatch=512`

### Steps to reproduce
1. Prepare one `embd.f32` file = `n_tokens × n_embd` little-endian float32 (random values are fine).
2. Build and run the minimal probe, mode `repeat`: it decodes the same batch N times in the **same context**, calling `llama_memory_clear(llama_get_memory(ctx), true)` before each decode, and reports `bitwise_equal` for each run vs run 0.
3. Also run mode `tokens` (same file interpreted as int32 tokens) as the control.

# Observed
| case | result |
| --- | --- |
| token path, 6 repeats | bitwise identical (cosine 1.00000000) |
| embd path, **1** token, 6 repeats | bitwise identical |
| embd path, **2/3/5** tokens, 6 repeats | **bitwise different** (cosine 0.984–0.9999; `argmax` constant) |
| embd path, all-zero input | all-zero output, bitwise identical |
| embd path, separate processes | different across processes (pairwise cosine 0.9946–0.9971) |

# Additional localization
- **Per-layer comparison** via `llama_get_hidden_state`: layers before the **first full-attention layer** are bitwise identical; divergence starts exactly at that layer (for this model: layers 0–2 identical, layer 3 differs). SSM layers are deterministic.
- **Requires >1 token**: 1 token is deterministic, 2 tokens already are not.
- Results fall into a **small set of states that recur** (e.g. run0==run2, run1==run3==run4), suggesting **buffer/state reuse** rather than fresh randomness.

# Ruled out
- unwritten `inp->tokens` (explicitly zeroing that input in `llm_graph_input_embd::set_input` does not help)
- multi-threading (`n_threads = n_threads_batch = 1`)
- KV capacity / unused KV region (swept `n_ctx` 8…512)
- reading uninitialized/stale KV content (all-zero input yields all-zero output)
- ggml fused ops (`GGML_CPU_DISABLE_FUSION=1`)
- CPU Flash Attention (`flash_attn_type = DISABLED`)

# Expected behavior
Same context + same input + cleared memory ⇒ **identical** output, as on the token path
(and as any user would expect for a pure-CPU deterministic backend).

# Extra context
This surfaced while prototyping **layer-pipeline relay** (compute layers `0..k` with one model instance and inject the resulting hidden state into a truncated model that computes `k+1..N` via `llama_batch.embd`). Greedy generation still matches the non-relay baseline 16/16 steps, but bit-exact reproducibility across processes is not achievable today.

### Relevant log output

<details>
<summary>Logs</summary>

```console

```
</details>

**Comment by mitoyosh** [comment]
Reproduced on Linux x86-64 (Debian 13, CPU build of master `930e2fa`,
`n_threads = n_threads_batch = 1`, flash attention off) with a public model,
`ggml-org/Qwen3.5-0.8B-GGUF` BF16, same probe shape as yours — and I think I
have the cause. It is a heap over-read of the caller's `pos` array, not
arithmetic.

**Symptom, same as yours** (plus: argmax does move here):

```
tokens N=5   run 1..5: bitwise_equal=yes  cosine=1.00000000  argmax_same=yes
embd   N=5   run 1: bitwise_equal=NO  max_abs_diff=2.4   cosine=0.99422627  argmax_same=yes
             run 2: bitwise_equal=NO  max_abs_diff=2.05  cosine=0.99525302  argmax_same=NO
             run 3: bitwise_equal=NO  max_abs_diff=2.11  cosine=0.99615401  argmax_same=yes
             run 4: bitwise_equal=NO  max_abs_diff=1.85  cosine=0.99524820  argmax_same=NO
             run 5: bitwise_equal=NO  max_abs_diff=2.06  cosine=0.99490347  argmax_same=yes
```

**Locating it.** Hashing every node's output through `cb_eval` on two decodes
of the same batch, the first node that differs (ignoring `node_0`, the
uncomputed token-path branch left by `ggml_build_forward_select`, which
nothing reads) is the RoPE on the first full-attention layer — matching your
per-layer result — and it differs because its *position input* differs:

```
#204   Qcur-3   op=ROPE   ne=[256,8,8,1]
        src1 leaf_55   DIFFERS from run 0   (128 bytes = int32 x 32 = 8 tokens x 4 sections)
```

**The read.** `llama-batch.cpp`, `ubatch_add`:

```cpp
for (size_t j = 0; j < (size_t)n_pos_per_embd; ++j) {
    // ... otherwise, the input batch is image embeddings, we copy the positions as-is
    size_t src_off = batch.token ? 0 : j*batch.n_tokens;
    udata->pos[j*n_tokens + i] = batch.pos[src_off + idxs[i]];
}
```

For an embd batch on an M-RoPE model this reads `batch.pos[0 .. 4*n_tokens)`.
But `llama.h` says "The provided arrays (i.e. token, embd, pos, etc.) must have
size of n_tokens", and `llama_batch_init` allocates `pos` with `n_tokens`
entries — so a caller doing what the header says (your probe, mine) hands over
N positions and sections 1–3 are read from the heap. valgrind on the unmodified
library, `embd N=5`:

```
Invalid read of size 4
   at llama_batch_allocr::ubatch_add(std::vector<int> const&, unsigned int, bool)
   by llama_batch_allocr::split_equal(unsigned int, bool, unsigned int)
   by llama_memory_hybrid::init_batch(llama_batch_allocr&, unsigned int, bool)
   by llama_context::decode(llama_batch const&)
 Address 0xb4e4244 is 0 bytes after a block of size 20 alloc'd
   at malloc
   by llama_batch_init
```

20 bytes = 5 positions. That also explains the things that looked like
variables and were not: which N fail, whether graph reuse or a fresh context
"fixes" it, and — the one I nearly published — weight type (the same file
converted to F16 or Q8_0 was deterministic; gemma-3-1b in BF16 was fine). All
of that is what happens to sit after a 20-byte malloc. `X_PERTURB_=1` in the
environment flipped the result too. Your zero-input control is stable because
garbage positions rotate zero to zero.

**Single-variable fix, caller side.** Same file, same build, `pos` sized
`4*N` with every section filled (`pos[j*N + i] = i`): every failing case is
bitwise deterministic across six runs, including the N that failed with
`LLAMA_GRAPH_REUSE_DISABLE=1` and with a fresh context per decode.

```
pos as documented (N entries):    N=3,5,8 NON-DETERMINISTIC
pos sized 4*N, all sections:      N=3,5,8,10 DETERMINISTIC   (also 7-10 with reuse off / fresh ctx)
```

So this is the `pos != NULL` twin of #28902 / #28910 (`pos == NULL`, where the
library's own buffer is sized `n_tokens` and read at `4*n_tokens`). Same loop,
same "embd on M-RoPE means image embeddings with four sections" assumption; here
the buffer is the caller's. Either the header should say that embd batches on
M-RoPE models carry `n_pos_per_embd * n_tokens` positions (and
`llama_batch_init` would need the model to size it), or the batch code should
broadcast when handed one position per token, as #28910 does for the NULL
case — with `mtmd`, which legitimately passes four sections, needing a way to
say so.

Probe (env `PROBE_POS4=1` applies the caller-side fix; `PROBE_FRESH=1`,
`PROBE_ZERO=1` as named):

<details>
<summary>probe.c</summary>

```c
// probe.c — llama.cpp#28963: is decoding through llama_batch.embd deterministic?
//
//   ./probe MODEL.gguf tokens N REPEATS [THREADS]
//   ./probe MODEL.gguf embd   N REPEATS [THREADS]
//
// Same context, memory cleared before every decode, identical input each time.
// Compares the full logits block (N positions x n_vocab) of every run against
// run 0: bitwise equal?  max |diff|?  cosine?  argmax same?
// embd input: deterministic pseudo-random floats (fixed seed), like the report.
//
// PROBE_FRESH=1  new llama_context per run
// PROBE_ZERO=1   all-zero embd input
// PROBE_POS4=1   allocate pos with n_pos_per_embd*N entries (4 on M-RoPE models)
//                and fill every section, instead of the N entries llama.h documents.
//                On an M-RoPE model the embd path reads pos[j*N + i] for j < 4
//                (llama-batch.cpp ubatch_add), i.e. past what llama_batch_init
//                allocated. This is the fix; without it the read is heap garbage.
#include "llama.h"
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static void quiet(enum ggml_log_level level, const char * text, void * ud) { (void)level; (void)text; (void)ud; }

int main(int argc, char ** argv) {
    if (argc < 5) { fprintf(stderr, "usage: %s model.gguf tokens|embd N REPEATS [THREADS]\n", argv[0]); return 2; }
    const char * model_path = argv[1];
    const int use_embd = strcmp(argv[2], "embd") == 0;
    const int N = atoi(argv[3]);
    const int R = atoi(argv[4]);
    const int T = argc > 5 ? atoi(argv[5]) : 1;
    const int FRESH = getenv("PROBE_FRESH") != NULL;   // new context per run
    const int ZERO  = getenv("PROBE_ZERO")  != NULL;   // all-zero embd input
    const int POS4  = getenv("PROBE_POS4")  != NULL;   // see header

    llama_log_set(quiet, NULL);
    llama_backend_init();

    struct llama_model_params mp = llama_model_default_params();
    mp.n_gpu_layers = 0;
    struct llama_model * model = llama_model_load_from_file(model_path, mp);
    if (!model) { fprintf(stderr, "load failed\n"); return 2; }
    const struct llama_vocab * vocab = llama_model_get_vocab(model);
    const int n_embd  = llama_model_n_embd_inp(model);
    const int n_vocab = llama_vocab_n_tokens(vocab);
    const enum llama_rope_type rt = llama_model_rope_type(model);
    const int n_pos = (rt == LLAMA_ROPE_TYPE_MROPE || rt == LLAMA_ROPE_TYPE_IMROPE) ? 4 : 1;

    struct llama_context_params cp = llama_context_default_params();
    cp.n_ctx = 512; cp.n_batch = 512; cp.n_ubatch = 512;
    cp.n_threads = T; cp.n_threads_batch = T;
    cp.flash_attn_type = LLAMA_FLASH_ATTN_TYPE_DISABLED;
    struct llama_context * ctx = llama_init_from_model(model, cp);
    if (!ctx) { fprintf(stderr, "ctx failed\n"); return 2; }

    // Fixed input. tokens: a real prompt; embd: seeded pseudo-random floats.
    llama_token toks[64];
    int n_tok = llama_tokenize(vocab, "The quick brown fox jumps over the lazy dog near the river bank", 63, toks, 64, 1, 0);
    if (!use_embd && n_tok < N) { fprintf(stderr, "prompt too short for N=%d\n", N); return 2; }

    struct llama_batch batch = llama_batch_init(N, use_embd ? n_embd : 0, 1);
    batch.n_tokens = N;
    if (POS4) { free(batch.pos); batch.pos = calloc((size_t)n_pos * N, sizeof(llama_pos)); for (int j = 0; j < n_pos; j++) for (int i = 0; i < N; i++) batch.pos[j * N + i] = i; }
    srand(12345);
    for (int i = 0; i < N; i++) {
        if (use_embd) {
            for (int k = 0; k < n_embd; k++) batch.embd[i * n_embd + k] = ZERO ? 0.0f : (float)rand() / RAND_MAX - 0.5f;
        } else {
            batch.token[i] = toks[i];
        }
        batch.pos[i] = i; batch.n_seq_id[i] = 1; batch.seq_id[i][0] = 0; batch.logits[i] = 1;
    }

    const size_t n_out = (size_t)N * n_vocab;
    float * ref = malloc(n_out * sizeof(float));
    float * cur = malloc(n_out * sizeof(float));

    printf("mode=%s N=%d repeats=%d threads=%d n_embd=%d n_vocab=%d rope_type=%d n_pos_per_embd=%d fresh=%d zero=%d pos4=%d\n", use_embd ? "embd" : "tokens", N, R, T, n_embd, n_vocab, (int)rt, n_pos, FRESH, ZERO, POS4);
    int all_bitwise = 1;
    for (int r = 0; r < R; r++) {
        if (FRESH && r > 0) { llama_free(ctx); ctx = llama_init_from_model(model, cp); }
        llama_memory_clear(llama_get_memory(ctx), true);
        if (llama_decode(ctx, batch) != 0) { fprintf(stderr, "decode failed run %d\n", r); return 2; }
        for (int i = 0; i < N; i++) memcpy(cur + (size_t)i * n_vocab, llama_get_logits_ith(ctx, i), n_vocab * sizeof(float));
        if (r == 0) { memcpy(ref, cur, n_out * sizeof(float)); printf("run 0: reference\n"); continue; }
        int bitwise = memcmp(ref, cur, n_out * sizeof(float)) == 0;
        double maxd = 0, dot = 0, na = 0, nb = 0; int argmax_same = 1;
        for (int i = 0; i < N; i++) {
            int a = 0, b = 0;
            for (int v = 0; v < n_vocab; v++) {
                size_t k = (size_t)i * n_vocab + v;
                double d = fabs((double)ref[k] - cur[k]); if (d > maxd) maxd = d;
                dot += (double)ref[k] * cur[k]; na += (double)ref[k] * ref[k]; nb += (double)cur[k] * cur[k];
                if (ref[k] > ref[(size_t)i * n_vocab + a]) a = v;
                if (cur[k] > cur[(size_t)i * n_vocab + b]) b = v;
            }
            if (a != b) argmax_same = 0;
        }
        printf("run %d: bitwise_equal=%s max_abs_diff=%.3g cosine=%.8f argmax_same=%s\n",
               r, bitwise ? "yes" : "NO", maxd, dot / sqrt(na * nb), argmax_same ? "yes" : "NO");
        if (!bitwise) all_bitwise = 0;
    }
    printf("RESULT %s N=%d: %s\n", use_embd ? "embd" : "tokens", N, all_bitwise ? "DETERMINISTIC" : "NON-DETERMINISTIC");
    llama_batch_free(batch); llama_free(ctx); llama_model_free(model); llama_backend_free();
    return all_bitwise ? 0 : 1;
}
```

</details>

Build: `gcc -O2 -o probe probe.c -Iinclude -Iggml/include -Lbuild/bin -lllama -lggml -lggml-base -lm -Wl,-rpath,build/bin`
Run: `./probe model.gguf embd 5 6 1` and `PROBE_POS4=1 ./probe model.gguf embd 5 6 1`.

Full write-up, including the false trail:
https://homelabpostmortem.com/2026/09/16/llama-cpp-reads-past-your-pos-array-for-embedding-batches-on-mrope-models/

**Comment by SgfKrc** [comment]
@mitoyosh 
Your diagnosis is correct, and it explains every observation we had collected. Thank you — I would not have found the `ubatch_add` over-read on my own.

Independent confirmation on a different platform / model / toolchain:

- **Platform**: Windows 11 x86-64, MSYS2 UCRT64 g++ 15.2.0, CPU backend (`n_gpu_layers = 0`), CMake "MinGW Makefiles" build of the hidden-state-extraction tree (`6f04274cc`).
- **Model**: Qwen3.5-2B (`qwen35`, f16 GGUF). Its `qwen35.rope.dimension_sections` has **4** entries, so I used `n_pos_per_embd = 4` as you described.
- **Caller-side fix applied exactly as reported**: for `embd` batches, allocate `pos` with `4 * n_tokens` entries and fill every section with the same positions (`pos[j*n_tokens + i] = i`).

Results before/after that single change (same binary otherwise, same inputs):

| case | before | after |
| --- | --- | --- |
| embd batch, same input, two runs, 141 positions | argmax 126/141, cosine 0.85–0.99 | **141/141, cosine min = 1.000000** |
| token path control, 141 positions | bitwise stable | bitwise stable (unchanged) |
| embd vs token path, same 141-token prompt | argmax 132/141 | **141/141, cosine min = 0.999999** |

I also hit exactly the "false variables" you listed: switching the weight type appeared to make it deterministic, a fresh context or disabled graph reuse appeared to change which N fails, and `X_PERTURB_` flipped results. My node-level observation that the first differing node is the RoPE on the first full-attention layer matches your finding.

One long-sequence data point that may be useful: in a layer-pipeline relay where a truncated model consumes the upstream hidden state through `llama_batch.embd`, a 141-step greedy generation diverged from the whole-model baseline at steps 46–67 (varying per run) **before** the fix. After applying the caller-side fix, the same 141-step relay matches the baseline exactly, repeatedly.

I am only confirming your report from my side — the analysis and the diagnosis are yours.
